lib/EnsEMBL/REST/Model/ga4gh/Beacon.pm

```diff
@@ -541,61 +553,77 @@ sub variant_exists {
-    if ($sv == 0 && ($seq_region_start != $new_start) && ($seq_region_end != $new_end)) {
+    # Precise match for snv or small indels - end does not make a difference
+    if ($sv == 0 && ($seq_region_start != $start_pos) && ($seq_region_end != $end_pos)) {
       next;
     }
     # Match for structural variants
     if($sv == 1){
-      next unless ($seq_region_start >= $new_start && $seq_region_start <= $start_max_pos && $seq_region_end >= $end_min_pos && $seq_region_end <= $new_end);
-    }
-
-    # All datasets where variant was found
-    my $datasets_found = $vf->get_all_VariationSets();
-    foreach my $set (@$datasets_found) {
-      $dataset_var_found{$set->dbID()} = $set;
+      next unless ($seq_region_start >= $start_pos && $seq_region_start <= $start_max_pos && $seq_region_end >= $end_min_pos && $seq_region_end <= $end_pos);
     }
 
     # Variant is a SNV
     if ($sv == 0) {
     $ref_allele_string = $vf->ref_allele_string();
     $alt_alleles = $vf->alt_alleles();
 
-    if ((uc($ref_allele_string) eq uc($new_ref))
-      && (grep(/^$new_alt$/i, @{$alt_alleles}))) {
+    # Some ins/del have allele string 'G/-', others have TG/T
+    # trim sequence doesn't always work
+    if ((uc($ref_allele_string) eq uc($ref_allele) || (uc($ref_allele_string) eq uc($new_ref)))
+      && (grep(/^$alt_allele$/i, @{$alt_alleles}) || grep(/^$new_alt$/i, @{$alt_alleles}))) {
         $found = 1;
+
+        # Checks datasets only for variants that match
+        my $datasets_found = $vf->get_all_VariationSets();
+        my @list_datasetids;
+        foreach my $set (@$datasets_found) {
+          my $dt_id = $set->dbID();
+          $dataset_var_found{$dt_id} = $set;
+          push (@list_datasetids, $dt_id);
+        }
+        $variant_dt{$vf->variation_name} = \@list_datasetids;
+
         if ($incl_ds_response) {
-          $vf_found = $vf;
+          push (@vf_found, $vf);
         }
-        last;
-      } else {
-          $found = 0;
       }
     }
     # Variant is a SV 
     else {
-      # convert to SO term 
+      # convert to SO term
+      # There's more than one term for each type - use a list of SO terms
+      # CNV =~ DEL or DUP 
       my %terms = (
-        INS  => 'insertion',
-        DEL  => 'deletion',
-        CNV  => 'copy_number_variation',
-        DUP  => 'duplication',
-        INV  => 'inversion',
+        'INS'    => 'insertion',
+        'INS:ME' => 'mobile_element_insertion',
+        'DEL'    => 'deletion',
+        'DEL:ME' => 'mobile_element_deletion',
+        'CNV'    => 'copy_number_variation,copy_number_gain,copy_number_loss,deletion,duplication',
+        'DUP'    => 'duplication',
+        'INV'    => 'inversion',
         'DUP:TANDEM' => 'tandem_duplication'
       );
 
       my $vf_so_term;
       $so_term = defined $terms{$alt_allele} ? $terms{$alt_allele} : $alt_allele;
       $vf_so_term = $vf->class_SO_term();
 
-      if ($vf_so_term eq $so_term) {
+      if ($so_term =~ /$vf_so_term/) {
```

Comment:
do we really want a regex here? CNV and contains deletion and duplication...
and please update the regex to use ${} for variables. it's now very confusing because you also use $ as start of string.

Reply:
regex deleted - using `contains_value` seems like a better idea.